// Copyright 2019 Tulip Solutions B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// [#protodoc-title: Private trade]

syntax = "proto3";
package tulipsolutions.api.priv;

option java_multiple_files = true;
option java_package = "nl.tulipsolutions.api.priv";
option go_package = "github.com/tulipsolutions/tecl/tulipsolutions/api/priv/trade";

import "validate/validate.proto";
import "tulipsolutions/api/common/orders.proto";
import "tulipsolutions/api/common/options.proto";

message GetPrivateTradesRequest {
    // Only return trades for provided markets. If empty, trades for all markets are returned.
    repeated common.Market markets = 1 [
        (validate.rules).repeated .unique = true,
        (validate.rules).repeated .items.enum = { not_in: [ 0 ], defined_only: true }
    ];
    common.SearchDirection search_direction = 2 [(validate.rules).enum = { not_in: [ 0 ], defined_only: true }];
    // The ID or timestamp of the earliest trade in the case of forward event search direction,
    // otherwise the ID or timestamp of the latest trade.
    // If omitted, defaults to the first or last order event, depending on the search direction.
    oneof start {
        fixed64 event_id = 3 [jstype = JS_STRING, (validate.rules).fixed64.gt = 0];
        fixed64 timestamp_ns = 4 [jstype = JS_STRING, (validate.rules).fixed64.gt = 0];
    }
    // Maximum number of trades returned. Defaults to 1000.
    uint32 limit = 5 [(validate.rules).uint32 = { gte: 0, lte: 1000 }, (common.default_limit) = 1000];
}

message StreamPrivateTradesRequest {
    // Only return trades for provided markets. If empty, trades for all markets are returned.
    repeated common.Market markets = 1 [
        (validate.rules).repeated .unique = true,
        (validate.rules).repeated .items.enum = { not_in: [ 0 ], defined_only: true }
    ];
    common.SearchDirection search_direction = 2 [(validate.rules).enum = { not_in: [ 0 ], defined_only: true }];
    // The ID or timestamp of the earliest trade in the case of forward search direction,
    // otherwise the ID or timestamp of the latest trade.
    // If omitted, ``start = now``
    oneof start {
        fixed64 event_id = 3 [jstype = JS_STRING, (validate.rules).fixed64.gt = 0];
        fixed64 timestamp_ns = 4 [jstype = JS_STRING, (validate.rules).fixed64.gt = 0];
    }
}

// Similar to :protobuf:message:`tulipsolutions.api.pub.PublicTrade`, but includes information about your matched Order
// ID and the fees charged.
//
// See also: :ref:`faq_trade_order_id_unique`
message PrivateTrade {
    fixed64 event_id = 1 [jstype = JS_STRING];
    common.Market market = 2;
    fixed64 order_id = 3 [jstype = JS_STRING];
    double price = 4;
    double base_amount = 5;
    double quote_amount = 6;
    double fee = 7;
    common.Currency fee_currency = 8;
    common.Side side = 9;
    fixed64 timestamp_ns = 10 [jstype = JS_STRING];
}

message PrivateTrades {
    repeated PrivateTrade trades = 1;
}

service PrivateTradeService {
    rpc GetTrades(GetPrivateTradesRequest) returns (PrivateTrades);
    rpc StreamTrades(StreamPrivateTradesRequest) returns (stream PrivateTrade);
}
